core/trino-main/src/main/java/io/trino/server/protocol/Query.java

```diff
@@ -450,6 +450,8 @@ private synchronized QueryResults getNextResult(long token, UriInfo uriInfo, Dat
         }
         else {
             nextToken = OptionalLong.empty();
+            // the client is not coming back, make sure the exchangeClient is closed
```

Comment:
Shouldn't this be handled above by `closeExchangeClientIfNecessary()`?

Reply:
`closeExchangeClientIfNecessary` only closes the client if a query fails or if a query returns no results (e.g.: a `DDL` query)